query/transformation.go

```diff
@@ -0,0 +1,195 @@
+// Copyright 2015 Square Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package query contains all the logic to parse
+// and execute queries against the underlying metric system.
+package query
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/square/metrics/api"
+)
+
+// A transform takes the list of values, other parameters, and the timerange of the query.
+type transform func([]float64, []value, api.Timerange) ([]float64, error)
+
+// timerangeScale takes an api.Timerange and computes the "scale" or duration of one sample (in seconds).
+// It is useful for transformations that normalize on time (like derivative or integral).
+func timerangeScale(timerange api.Timerange) float64 {
+	return float64(timerange.Resolution)
+}
+
+// transformTimeseries transforms an individual series (rather than an entire serieslist) taking the same parameters as a transform,
+// but with the serieslist standing in for the simplified []float64 argument.
+func transformTimeseries(series api.Timeseries, transform transform, parameters []value, timerange api.Timerange) (api.Timeseries, error) {
+	values, err := transform(series.Values, parameters, timerange)
+	if err != nil {
+		return api.Timeseries{}, err
+	}
+	return api.Timeseries{
+		Values: values,
+		TagSet: series.TagSet,
+	}, nil
+}
+
+// applyTransform applies the given transform to the entire list of series.
+func ApplyTransform(list api.SeriesList, transform transform, parameters []value) (api.SeriesList, error) {
+	result := api.SeriesList{
+		Series:    make([]api.Timeseries, len(list.Series)),
+		Timerange: list.Timerange,
+		Name:      list.Name,
+	}
+	var err error
+	for i, series := range list.Series {
+		result.Series[i], err = transformTimeseries(series, transform, parameters, list.Timerange)
+		if err != nil {
+			return api.SeriesList{}, err
+		}
+	}
+	return result, nil
+}
+
+// checkParameters is used to make sure that each transform is given the right number of parameters.
+func checkParameters(name string, expected int, parameters []value) error {
+	if len(parameters) != expected {
+		printArgs := append([]value{stringValue("(SeriesList)")}, parameters...)
```

Comment:
Transformation functions are given `[]float64` corresponding to a single timeseries (from a serieslist), in addition to a `parameters []value` which corresponds to all the *other* parameters given in the query. For example, when writing:

`transform.moving_average( cpu, 60 )`

The `parameters` will be the slice `[]value{ 60 }` (ignoring type wrappers).

However, I wanted to try to hide this from the user (since they don't need to care how it's implemented) and they'd like to see the parameters including the time series. The `append` puts a string that says `"(SeriesList)"` to stand for the currently-processing series-list which isn't actually available to the function. Otherwise, when, say, they call `transform.derivative` accidentally with a second argument, it would only show them the second argument.

There might be a better way of doing it, but I'm not familiar with the Go way.